components/ord-service/src/main/java/com/sap/cloud/cmp/ord/service/token/Token.java

```diff
@@ -40,23 +41,42 @@ public class Token {
 
     private Set<String> formationIDsClaims;
 
-    public Token(SubscriptionHelper subscriptionHelper, String idTokenEncoded) throws JsonMappingException, JsonProcessingException {
+    public String applicationTenantId;
+
+    public Token(SubscriptionHelper subscriptionHelper, String idTokenEncoded, String applicationTenantId) throws JsonMappingException, JsonProcessingException {
         this.subscriptionHelper = subscriptionHelper;
         String idTokenDecoded = decodeIDToken(idTokenEncoded);
         this.content = mapper.readTree(idTokenDecoded);
         this.formationIDsClaims = new HashSet<>();
+        this.applicationTenantId = applicationTenantId;
     }
 
     public String extractTenant() {
         String unescapedTenants = content.get(TENANTS_MAP_KEY).asText().replace("\\", "");
 
         String providerTenantID = "";
         try {
+            SelfRegisteredRepository repo = subscriptionHelper.getRepo();
+
             JsonNode tenantsTree = mapper.readTree(unescapedTenants);
 
             String tenant = tenantsTree.path(CONSUMER_TENANT_KEY).asText();
             providerTenantID = tenantsTree.path(PROVIDER_TENANT_KEY).asText();
             if (providerTenantID == null || providerTenantID.isEmpty() || providerTenantID.equals(tenant)) {
+                if (applicationTenantId != null && !applicationTenantId.isEmpty()) {
+                    logger.info("Application local tenant ID is provided through header");
+                    String consumerID = content.get(CONSUMER_ID_KEY).asText();
+                    if (consumerID == null || consumerID.isEmpty()) {
+                        logger.error("consumer ID could not be empty");
+                        return tenant;
+                    }
+                    logger.info("Checking for application with local tenant ID {} and app template ID {}", applicationTenantId, consumerID);
+                    String appId = repo.findApplicationByLocalTenantIdAndApplicationTemplateId(applicationTenantId, consumerID);
```

Comment:
We discussed it offline but will leave the comment for tracking purposes.

Once we get the application, we should check if it has `global_subaccount_id` label and if so to get the internal id of that subaccount and return it. This is needed so that later in the ODataController we can set the subaccount id as `destination_tenant_id` claim allowing us to discover destinations.

This change/adaptation can be done in a separate follow-up PR.

Reply:
It will be addressed in a separate PR